pkg/monitors/http/http.go

```diff
@@ -0,0 +1,179 @@
+	cancel func()
+	logger logrus.FieldLogger
+}
+
+// Configure and kick off internal metric collection
+func (m *Monitor) Configure(conf *Config) error {
+	m.logger = logrus.WithFields(logrus.Fields{"monitorType": monitorType})
+	// Start the metric gathering process here
+	var ctx context.Context
+	ctx, m.cancel = context.WithCancel(context.Background())
+	utils.RunOnInterval(ctx, func() {
+		tlsCfg := &tls.Config{InsecureSkipVerify: true}
+		timeoutDuration := time.Duration(conf.Timeout) * time.Second
+		// add data
+		var body io.Reader
+		if conf.Body != "" {
+			body = strings.NewReader(conf.Body)
+		}
+		// get stats for each website
+		for _, site := range conf.URLs {
+			m.logger = m.logger.WithFields(logrus.Fields{"url": site})
+			m.logger.Debug("starting monitor url")
+			var dps []*datapoint.Datapoint
+			dimensions := map[string]string{"url": site}
+			dialer := &net.Dialer{Timeout: timeoutDuration}
+			client := &http.Client{
+				Transport: &http.Transport{
+					DialContext:       dialer.DialContext,
+					DisableKeepAlives: true,
+					TLSClientConfig:   tlsCfg,
+				},
+				Timeout: timeoutDuration,
+			}
+			if !conf.FollowRedirects {
+				client.CheckRedirect = func(req *http.Request, via []*http.Request) error {
+					return http.ErrUseLastResponse
+				}
+			}
+			req, err := http.NewRequest(conf.Method, site, body)
+			if err != nil {
+				m.logger.WithError(err).Error("could not create new request")
+				continue
+			}
+			// add hearders
+			for key, val := range conf.Headers {
+				req.Header.Add(key, val)
+				if key == "Host" {
+					req.Host = val
+				}
+			}
+			// starts timer
+			now := time.Now()
+			resp, err := client.Do(req)
+			if err != nil {
+				m.logger.WithError(err).Error("could not do the request")
+				continue
+			}
+			dps = append(dps, datapoint.New("http.response_time", dimensions, datapoint.NewFloatValue(time.Since(now).Seconds()), datapoint.Gauge, time.Time{}))
+			dps = append(dps, datapoint.New("http.status_code", dimensions, datapoint.NewIntValue(int64(resp.StatusCode)), datapoint.Gauge, time.Time{}))
+			defer resp.Body.Close()
+			bodyBytes, err := ioutil.ReadAll(resp.Body)
+			dps = append(dps, datapoint.New("http.content_length", dimensions, datapoint.NewIntValue(int64(len(bodyBytes))), datapoint.Gauge, time.Time{}))
+			if err != nil {
+				m.logger.WithError(err).Error("could not parse body")
+			} else if conf.Regex != "" {
+				var match int64 = 0
+				regex, err := regexp.Compile(conf.Regex)
+				if err != nil {
+					m.logger.WithError(err).Error("failed to compile regular expression")
+				}
+				if regex.Match(bodyBytes) {
+					match = 1
+				}
+				dps = append(dps, datapoint.New("http.regex_match", dimensions, datapoint.NewIntValue(match), datapoint.Gauge, time.Time{}))
+			}
+			parsedURL, err := url.Parse(resp.Request.URL.String())
+			if err != nil {
+				m.logger.WithError(err).Error("could not parse url")
+				m.Output.SendDatapoints(dps...)
+				continue
```

Comment:
If all the time datapoints that were gathered should still be sent (appears to be all the time?) then should refactor sending to a central place instead of relying on multiple call sites to.

Reply:
I updated the doc. at the origin, there are dedicated metrics like tls valid or regex match and should be reported only if applicable but now it is in dimensions so this is only true for tls expiration metric which will be reported only if https could connect